<a href="https://colab.research.google.com/github/cboyda/MachineLearning/blob/main/PA1_Team1_W23_ipynd.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PA1: Problem Scoping**

Team member names:

*  Brett Adams
*  Cailenys Leslie
*  Clinton Boyda
*  Tanvir Hossain
*  Ram Dershan

Dataset: 
[New York City Airbnb Open Data](https://www.kaggle.com/datasets/dgomonov/new-york-city-airbnb-open-data)

Where did all the kaggle code go?  Please do NOT delete group ideas/code until we have agreed on the final result together, we want to encourage alterative suggestions... [Clinton]

In [ ]:
import numpy as np
import pandas as pd
import plotly.express as px

In [7]:
#Inconsistent Google drive folders and paths necessitates leveraging single public url for data source, chose GitHub

filename = "https://raw.githubusercontent.com/cboyda/MachineLearning/main/AB_NYC_2019.csv"

df = pd.read_csv(filename)
df

id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48890        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891        Marisol            Brooklyn            Bushwick  40.70184   
48892  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893            Taz           Manhattan      Hell's Kitchen  40.75751   
48894     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaN                NaN                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
0                   365  
1                   355  
2                   365  
3                   194  
4                     0  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[48895 rows x 16 columns]

In [8]:
df.shape

(48895, 16)

In [9]:
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

In [10]:
label = "price"

In [11]:
df.describe()

id       host_id      latitude     longitude         price  \
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000   
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687   
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000   
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000   
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48895.000000       48895.000000       38843.000000   
mean         7.029962          23.274466           1.373221   
std         20.510550          44.550582           1.680442   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

# Group Discussion
## Background Group Consensus on Data seen:
Column details:
* A.   id = random unique rowID
* B.   name = name of add (unique?)
* C.   host_id = who the host is (not_unique since one host could have multiple listings = assumed)
* D:   host_name = human name of host (not_unique since one host could have multiple listings = assumed)
* E:   neighbourhood_group = name of general area/location (not unique)
* F:   neighbourhood = name of specific location in neighbourhood group (not unique)  
* G:   latitude = lat float for geo-location (not unique)
* H:   longitude = long float for geo-location (not unique)
* I:   room_type = category of type of rental (not unique)
* J:   price = price of room per night as integer? (not unique)
* K:   minimum_nights = host entered value for min staff **(necessary??)**
* L:   number_of_reviews = integer count of # of reviews (not unique) (missing usefulness if we don't know the actual rating or average rating this just seems like a count <> **necessary??**)
* M:   last_review = date of last review **(necessary??)**
* N:   reviews_per_month = float, but we don't know if this correlates with number of nights rented [are renters forced to complete a review?] (**necessary??**)
* O:   calculated_host_listings_count = what is this counting? # of rentals for specific host?
* P:   availability_365 = what is this counting? suspect range is 0 to 365 but what does 194 mean, that this was only available for rent 194 days of the year?

MISSING
* profit = have no clue what profit is (since we only know retail price)
* num_rents = have no clue which is most popular rental (count of actual rentals of unit seems missing)
* total_revenue = profit * num_rents

Do we start making assumptions about the data?
* num_rents = number_of_reviews ? but then look at row 4, min nights is 10, number of reviews = 9 ??
* all we can say is popularity HIGH if # of reviews is high
* if we are NOT using a distance calculation we could drop lat/long




In [12]:
df.loc[4]

id                                                                            5022
name                              Entire Apt: Spacious Studio/Loft by central park
host_id                                                                       7192
host_name                                                                    Laura
neighbourhood_group                                                      Manhattan
neighbourhood                                                          East Harlem
latitude                                                                  40.79851
longitude                                                                -73.94399
room_type                                                          Entire home/apt
price                                                                           80
minimum_nights                                                                  10
number_of_reviews                                                                9
last

## **Problem Scoping: Backwards Process**

1. **Outcome:** *Predict the annual revenue of a rental property on Airbnb in New York City*

> This project will result in a machine learning model that can accurately predict the annual revenue of a rental property on Airbnb in New York City. The model will be a valuable tool for property owners and managers to help them optimize their pricing strategies and maximize their return on investment.

Analysis that better matches data provided:
* Option A: Given a [**location**] and type of rental, predict the price of the Airbnb property in New York City
* Option B: Given a specific GPS location (lat/long) find/predict the distance to the min/average/max price of a property within x meters from that location.
* Option C: Given a [**location**] find the closest most popular location and return that price.  Where the assumption is high popularity = high number of reviews.  Prediction would be on 'popularity' based on **location** and price.

[**location**] can either be neighbourhood name, neighbourhood group name, or GPS location (lat/long), if we want to calculate distance.

QUESTION: Which location do we want to use?
* neighbourhood (easy)
* neighbourhood group (easy)
* GPS location (lat/long) and calculate distance (more challenging)


Comment (Ram): This outcome seems too broad to me (unless you meant price instead of annual revenue). Also, annual revenue is not a field/column we have in the data and unfortunately the data doesn't give us enough info to calculate annual revenue, so imo annual revenue is not a good label for our model. However, instead of annual revenue, I would suggest predicting price. So then like the second half of the paragraph mentions, the model would be a tool to help property owners set competitive prices on their properties.*

comment(Tanvir): I agree with Ram. Since we do not have profit data we cannot be able to provide revenue for a property. But we can do one thing in terms of business that makes a classification based on the neighbourhood (area, and parking are important for a neighbourhood), room type (shared with whom is a matter of security) and price. We can do two models one is regression-based as Ram mentioned, and another classification based to provide a complete scenario for a business owner or user. Airbnb owners can use the price prediction and users can use the classification model. It will be interesting to classify the model based on longitude and latitude to see the more tourist favourable area to stay.

Comment (Clinton): I say we start with option A since it nicely correlates with values we already have.  If we have more time we could do more complex distance calculations but that may be a distraction from the ML purpose of this assignment and take too much time away from the actual process.

Comment (Cailenys): I see what you are saying guys. But I don’t see how we can predict the price either. There are many factors to consider to predict the price, including the demand and historical data of the price per property. Are we going to assume those? 

Looking at the data, what I think we can do is to calculate the average rental price by neighbourhood and type of house, and then we can compare all the neighbourhoods to understand the market in the city. Although, I don’t know if we are going to add real value with this.   


2. **Performance, time and cost approximation**

Comment (Clinton): Since a pulldown of locations is rather boring, we could make Option A slightly more complicated by having the image https://drive.google.com/drive/folders/1w_KitUbWXom0AyPebb9Hv6ZYL_lbF_K5 be clickable?

Since there are only ~49,000 rows and 16 features, performance should not be a concern for processing.

Creation time (therefore costs) will be dependant on data cleanup required and how graphically enhanced we want to make the interface that queries the user.  As a iterative approach to maximize functional deliverables we will start with something trivial like drop down lists and move to interactive New York map selections as time and resources allow. 

For example utilizing Google maps or OpenStreet Data for clickable method of requesting a users location will require additional development overhead.

The key will be how much work can be done and still meet our weekly goals.


3. **Value Area:** *Project*

> The objective of this project is to develop a machine learning model that can predict the annual revenue of a rental property on Airbnb. The model will be trained on historical data and will use various factors such as location, property type, number of rooms, guest reviews, and other relevant factors to make accurate predictions.

Comment (Ram): We can discuss this on Thursday, but I feel that guest reviews is not a good feature for our prediction model. Considering that not all properties in the data have reviews and that in context not every tenant would leave a review, makes guest reviews an unreliable feature.*

Comment (Tanvir): We can propose a model that could be valuable for both sides (owner and user) since there are no restrictions. we can discuss more later.

Comment (Clinton): needs to include comments regarding (Prevent, Project, Produce, Promote, Personalize) your business outcome is related to.
Alternative value proposition...

> The result will product a pricing option once a location and type of room is provided.  This would promote services available by area for either future hosts or travellers.  To further add more value pricing could be statiscally displayed as min/avg/max to help show the options available.  

> Future analysis could then be made against other data sources to correlate crime with pricing information to measure prevention and progress of mitigation strategies, assuming as crime decreases prices would increase.

4. **Potential Actions:**

> *   Preprocess and clean the chosen dataset to ensure it is suitable for training the machine learning model.
> *   Select the most relevant features for the model. This will involve analyzing the data to determine which features impact the annual revenue of a rental property.
>* Label the New York neighbourhood map as a legend for our location selection
>*  Consider methods for finding a users current location (GPS) if we wish to complete any distance calcuations.


5. **Judgement Analysis:**

> Assuming Option A selected...
* A right prediction will return the value of Airbnb unit available with the consequence being supplying the correct information to the user.
* A wrong prediction is not critical or life altering because it only returns an erroneous price.
* Since the danger of a poor result is not life-threatening returning a single value (right or wrong) seems acceptable but could be further protected by returning a range of values (for example instead of 80 we could return 60 to 120 where these values are the min/max). Or further statistical analysis could be used to return a range of the highest frequency instead of the absolute min/max.
* A wrong result could also be returned if the pricing is provided for units that are not available.  While we know the number of days not available, we don't know specifically when those dates are so we can not specify this exact response unless the specific dates not available were provided (and the user asked for dates of travel).

6. **ML Answers available:**

> Assuming Option A selected...
* A right prediction will return the value of Airbnb unit in that area.

7. **Business Problem Context:**

> Assuming Option A selected...
* Business problem is to find the price of Airbnb listings given the location and type of room desired.

8. **Ethical Problem Description:**

> Assuming Option A selected...
* Huge biases exist that are missing from this data including cost of living, crime, and other factors that often correlate with property values.  All these are missing from this current data.
* Only having data from one year is missing any trend information.  With global changes to businesses due to things like pandemics we would need to see multiple years of information to truly know if these rates are representative.
